src/main/java/org/dbos/apiary/mysql/MysqlConnection.java

```diff
@@ -0,0 +1,186 @@
+        MysqlContext ctxt = new MysqlContext(this.connection.get(), workerContext, txc, service, execID, functionID);
+        FunctionOutput f = null;
+        try {
+            f = workerContext.getFunction(functionName).apiaryRunFunction(ctxt, inputs);
+        } catch (Exception e) {
+            e.printStackTrace();
+        }
+        return f;
+    }
+
+    @Override
+    public void rollback(Map<String, List<String>> writtenKeys, TransactionContext txc) {
+        String query = "";
+        try {
+            Connection c = ds.getConnection();
+            Statement s = c.createStatement();
+            for (String table : writtenKeys.keySet()) {
+                query = String.format("DELETE FROM %s WHERE %s = %d", table, MysqlContext.beginVersion, txc.txID);
+                s.execute(query);
+            }
+            s.close();
+            c.close();
+        } catch (Exception e) {
+            e.printStackTrace();
+            logger.error("Failed to rollback txn {}", txc.txID);
+            logger.info("Rollback query: {}", query);
+        }
+    }
+
+    @Override
+    public boolean validate(Map<String, List<String>> writtenKeys, TransactionContext txc) {
+        Set<Long> activeTransactions = new HashSet<>(txc.activeTransactions);
+        validationLock.lock();
+        boolean valid = true;
+        for (String table: writtenKeys.keySet()) {
+            for (String key : writtenKeys.get(table)) {
+                // Has the key been modified by a transaction not in the snapshot?
+                Set<Long> writes = committedWrites.getOrDefault(table, Collections.emptyMap()).getOrDefault(key, Collections.emptySet());
+                for (Long write : writes) {
+                    if (write >= txc.xmax || activeTransactions.contains(write)) {
+                        valid = false;
+                        break;
+                    }
+                }
+            }
+        }
+        if (valid) {
+            for (String collection: writtenKeys.keySet()) {
+                for (String key : writtenKeys.get(collection)) {
+                    committedWrites.putIfAbsent(collection, new ConcurrentHashMap<>());
+                    committedWrites.get(collection).putIfAbsent(key, ConcurrentHashMap.newKeySet());
+                    committedWrites.get(collection).get(key).add(txc.txID);
+                }
+            }
+        }
+        validationLock.unlock();
+        if (valid) {
+            String query = "";
+            try {
+                Connection c = ds.getConnection();
+                Statement s = c.createStatement();
+                for (String table : writtenKeys.keySet()) {
+                    for (String key : writtenKeys.get(table)) {
+                        query = String.format("UPDATE %s SET %s = %d WHERE %s = '%s' AND %s < %d AND %s = %d", table, MysqlContext.endVersion, txc.txID, MysqlContext.apiaryID, key, MysqlContext.beginVersion, txc.txID, MysqlContext.endVersion, Long.MAX_VALUE);
```

Comment:
Is it better to assemble multiple statements and execute them at once to reduce # round trips?

Reply:
Updated this part to make the inner loop a single SQL query.